app/services/session_encryptor.rb

```diff
@@ -1,35 +1,48 @@
 class SessionEncryptor
   def initialize
-    @user_access_key = reload_user_access_key
-  end
-
-  def duped_user_access_key
-    # Reload if UserAccessKey constant has been reloaded
-    @user_access_key = reload_user_access_key unless @user_access_key.is_a? UserAccessKey
-
-    # Return a clone since encryptor.decrypt mutates this key
-    @user_access_key.dup
+    @initialized_user_access_key = nil
+    @encryption_user_access_key = nil
+    @decryption_user_access_key_map = {}
   end
 
   def load(value)
-    decrypted = encryptor.decrypt(value, duped_user_access_key)
-
+    user_access_key = decryption_user_access_key(value)
+    decrypted = encryptor.decrypt(value, user_access_key)
     JSON.parse(decrypted, quirks_mode: true).with_indifferent_access
   end
 
   def dump(value)
     plain = JSON.generate(value, quirks_mode: true)
-    encryptor.encrypt(plain, duped_user_access_key)
+    encryptor.encrypt(plain, encryption_user_access_key)
+  end
+
+  def initialized_user_access_key
+    # Initialize access key once and share it since scrypt hashes take a long time to calculate
+    # Return a dupe since encryption operations mutate this key
+    return @initialized_user_access_key.dup if @initialized_user_access_key.is_a?(UserAccessKey)
+
+    key = Figaro.env.session_encryption_key
+    @initialized_user_access_key = UserAccessKey.new(password: key, salt: key)
   end
 
   private
 
-  def encryptor
-    Pii::PasswordEncryptor.new
+  attr_reader :decryption_user_access_key_map
+
+  def decryption_user_access_key(encrypted_value)
+    encryption_key = encrypted_value.split('.').first
```

Comment:
Is it guaranteed that the encrypted_value can be split at a period and that the key will be the first element in that array?

Reply:
Yeah, you can see where the encrypted blob is created here: https://github.com/18F/identity-idp/blob/master/app/services/pii/password_encryptor.rb#L11

The `join_segments` method is defined here: https://github.com/18F/identity-idp/blob/master/app/services/pii/encryptor.rb#L52